### DISCUSSION TITLE
How does `onAuthStateChange` work?

### DISCUSSION THREAD

**Discussion by gabedealmeida** [supabase/supabase#3673] [Questions] [answered]
Hey team!

I've been reading the docs and I'm very interested in understanding how `onAuthStateChange` works under the hood. How does it receive the notifications? Is it through WebSockets? Or are these notifications all just generated client-side in response to client initiated events, i.e. a client logs in, a client logs out. If it's the latter, how is JWT expiration handled? Is an event fired for that? What does that event look like, what's it called? Thanks, team!

**Comment by dshukertjr** [comment, marked as answer]
Hi @gabedealmeida!

`onAuthStateChange` events are generated client-side. You can see an example event being generated [here](https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L151). At the moment, there is no dedicated event for when the JWT expires, but you will receive `'SIGNED_IN'` as you can see [here](https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L619).

You can also take a look at all of the available events [here](https://github.com/supabase/gotrue-js/blob/master/src/lib/types.ts#L15).
